### PR TITLE
Support semver version ranges for `released` decorator (and manifests)

### PR DESCRIPTION
**Pull request by simon-id**
## Motivation

In Node.js we have multiple release lines (ie: v3.x, v4.x, etc...). Historically we would put the latest version in the system-tests manifest, without specifying the potentially backported versions.

Example:
Feature `Test_Blocking_request_cookies` has been released in v3.37.0 and v4.16.0.
Current manifest:
```yaml
Test_Blocking_request_cookies: v4.16.0
```

Ideally we'd want to specify the full compatibility range like this:
```yaml
Test_Blocking_request_cookies: ^3.37.0 || >=4.16.0
```
Where `^3.37.0` is equivalent to `>=3.37.0 <4.0.0`

## Changes

Yeah basically you can do it now, thanks to the pip package `semantic_version` and yours truly.
I kept it compatible with the current `vX.X.X` syntax, so if the version declaration starts with `v`, it will use the classic "above or equal", else it will consider it as a semver spec.

## TODO
- [x] Update documentation
- [x] Update manifest JSON Schema
- [x] Add tests ?

## Workflow

1. ⚠️ Create your PR as draft ⚠️
2. Work on you PR until the CI passes (if something not related to your task is failing, you can ignore it)
3. Mark it as ready for review
    * Test logic is modified? -> Get a review from RFC owner. We're working on refining the `codeowners` file quickly.
    * Framework is modified, or non obvious usage of it -> get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)

:rocket: Once your PR is reviewed, you can merge it!

🛟 [#apm-shared-testing](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X) 🛟

## Reviewer checklist

* [ ] [Relevant label](https://github.com/DataDog/system-tests/blob/main/docs/CI/labels.md) (`run-parametric-scenario`, `run-profiling-scenario`...) are presents
* [ ] No system-tests internal is modified. Otherwise, I have the approval from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
* [ ] CI is green, or failing jobs are not related to this change (and you are 100% sure about this statement)
* [ ] A docker base image is modified?
    * [ ] the relevant `build-XXX-image` label is present
    * [ ] To R&P team: locally build and push the image to hub.docker.com 
* [ ] A scenario is added (or removed)?
    * [ ] Get a review from [R&P team](https://dd.enterprise.slack.com/archives/C025TJ4RZ8X)
    * [ ] Once merged, add (or remove) it in system-test-dasboard nightly

### CODE REVIEW

**Comment by simon-id** [comment]
I tested all the nominal cases manually and it works perfectly. But I had enough python today to write unit tests for this, please someone from the @DataDog/system-tests-core team take over the PR.

**Review by uurien — APPROVED** [review]

File: docs/edit/manifest.md [hunk 1]
```diff
@@ -13,10 +13,14 @@ tests/:
...
 
       Test_FeatureC: # declare a version for a class, depending on weblog
-        '*'': missing_feature # All other weblogs: not yet available
+        '*': missing_feature # All other weblogs: not yet available
         django: v1.2
         flask: v1.3
         uwsgi: bug (jira ticket) # For a weblog, skip it with bug, or flaky
+      
+       # declare compatibility for multiple release lines
+       # the caret character locks the major version
```
Comment (simon-id) [inline docs/edit/manifest.md:22]:
```suggestion
       # the caret character locks the major version (ie: `>=1.3.0 && <2.0.0`)
```

File: utils/_decorators.py [hunk 2]
```diff
@@ -187,8 +188,12 @@ def compute_declaration(only_for_library, component_name, declaration, tested_ve
                 return declaration
 
             # declaration must be now a version number
-            if tested_version >= declaration:
-                return None
+            if declaration.startswith("v"):
+                if tested_version >= declaration:
+                    return None
+            else:
+                if semver.Version(str(tested_version)) in semver.NpmSpec(declaration):
```
Comment (simon-id) [inline utils/_decorators.py:195]: `semver.Version()` will throw if test_version is not a valid semver, should we put a try catch ?

File: tests/test_the_test/test_manifest.py [hunk 3]
```diff
@@ -81,6 +81,3 @@ def assert_in(elements, module, nodeid):
 
             for variant in declaration:
                 assert variant in variants_map[component], f"Variant {variant} does not exists for {component}"
-
-
-test_content()
```
Comment (simon-id) [inline tests/test_the_test/test_manifest.py:86]: did you skip the manifest tests ? why

**Comment by simon-id** [comment]
The 2 python parametric jobs failing are also failing in master :shrug:

**Review by smola — APPROVED** [review]

**Comment by simon-id** [comment]
@DataDog/system-tests-core

**Comment by simon-id** [comment]
I am not sure we can easily reconcile this SemVer spec with the manifest schema in `/manifests/parser/schema.json`. Maybe the manifest validation should be done python side instead ?

**Comment by cbeauchesne** [comment]
`./format.sh` to fis the lint job

File: tests/test_the_test/test_manifest.py (hunk 3 above)
Comment (cbeauchesne) [reply tests/test_the_test/test_manifest.py:86]: put it back please

**Review by cbeauchesne — CHANGES_REQUESTED** [review]

**Review by cbeauchesne — APPROVED** [review]

**Pull request merged**